File: 2022/day-2.c
#include <stdio.h>

int main(){
    FILE *fp = fopen("input.txt", "r");

    int total = 0;
    while(!feof(fp)){
        char a, b;
        fscanf(fp, "%c %c\n", &a, &b);

        if(a - 'A' == b - 'X'){
            total += 3;
        }
        else if(a == 'A' && b == 'Y' ||
                a == 'B' && b == 'Z' ||
                a == 'C' && b == 'X'){
            total += 6;
        }

        total += b - 'X' + 1;
    }

    printf("%d\n", total);
    return 0;
}

File: 2025/day-2.js
var fs = require('fs'),
    path = require('path'),    
    filePath = path.join(__dirname, 'input.txt');

const isInvalid = (id) => {
    const strId = String(id.toString());
    if (strId.length % 2 !== 0) {
        return false;
    }
    return strId.substring(0, strId.length / 2) === strId.substring(strId.length / 2);
}

let total = 0;
fs.readFile(filePath, {encoding: 'ascii'}, function(err,data){
    if (!err) {
        let datas = data.trim().split(',');
        console.log(datas);
        for (const ranges in datas) {
            let range = datas[ranges].split('-');
            let [first, second] = range.map(x => parseInt(x));
            const delta = second - first;
            if (first.toString().length === second.toString().length && first.toString().length % 2 !== 0) {
                continue;
            }
            else if (first.toString().length !== second.toString().length) { // TODO: optimize this logic
                for (let id = first; id <= second; id++) {
                    if (isInvalid(id)) {
                        total += id;
                    }
                }
            }
            else {
                for (let id = first; id <= second; id++) { // TODO: optimize this logic
                    if (isInvalid(id)) {
                        total += id;
                    }
                }
            }
        }        

        console.log('Sum of invalid ids:', total);
    } else {
        console.log(err);
    }
});
